File: 712.py
class Solution:
    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        dp = [0] * len(s2)
        dp[0] = 0 if s1[0] == s2[0] else (ord(s1[0]) + ord(s2[0]))
        prefixSum = ord(s2[0])
        for j in range(1, len(s2)):
            dp[j] = prefixSum if s1[0] == s2[j] else (dp[j - 1] + ord(s2[j]))
            prefixSum += ord(s2[j])
        prefixSum = ord(s1[0])
        for i in range(1, len(s1)):
            topleft = dp[0]
            dp[0] = prefixSum if s1[i] == s2[0] else (dp[0] + ord(s1[i]))
            prefixSum += ord(s1[i])
            for j in range(1, len(s2)):
                topleft, dp[j] = dp[j], (topleft if s1[i] == s2[j] else min(dp[j - 1] + ord(s2[j]), dp[j] + ord(s1[i])))
        return dp[-1]

File: sol/712/712.cpp
struct Solution {
    int minimumDeleteSum(const string &s1, const string &s2) {
        vector<int> dp(s2.size(), 0);
        int prefixSum = s2[0], topleft, temp;
        dp[0] = (s1[0] == s2[0] ? 0 : s1[0] + s2[0]);
        for (int j = 1; j < s2.size(); ++j){
            dp[j] = (s1[0] == s2[j] ? prefixSum : dp[j - 1] + s2[j]);
            prefixSum += s2[j];
        }
        prefixSum = s1[0];
        for (int i = 1; i < s1.size(); ++i){
            topleft = dp[0];
            dp[0] = (s1[i] == s2[0] ? prefixSum : dp[0] + s1[i]);
            prefixSum += s1[i];
            for (int j = 1; j < s2.size(); ++j){
                temp = topleft;
                topleft = dp[j];
                dp[j] = (s1[i] == s2[j] ? temp : min(dp[j - 1] + s2[j], dp[j] + s1[i]));
            }
        }
        return dp.back();
    }
};
